### PR TITLE
[hail] Refactor getOrDefineMethod to be on FB, not EFB

### PR DESCRIPTION
**Pull request by tpoterba**
Stacked on #8099

### CODE REVIEW

File: hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala [hunk 1]
```diff
@@ -328,7 +343,7 @@ class FunctionBuilder[F >: Null](val parameterTypeInfo: Array[MaybeGenericTypeIn
       cn.accept(cw)
       bytes = cw.toByteArray
 //       This next line should always be commented out!
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala]: Can we get rid of this comment?

Comment (tpoterba) [reply hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala]: no, the line below should be commented out! This is fixed in the parent PR (rvb copy)

**Comment by tpoterba** [comment]
(don't review this until the parent is in!)

**Comment by akotlar** [comment]
this looks ready to go in once the rvb pr is merged

**Comment by akotlar** [comment]
rebase and then should be good to go!

**Comment by tpoterba** [comment]
done!

**Pull request merged**